src/Common/RWLock.cpp

```diff
@@ -278,42 +302,78 @@ void RWLockImpl::dropOwnerGroupAndPassOwnership(GroupsContainer::iterator group_
         readers_queue.erase(group_it);
         /// Prepare next phase
         if (!writers_queue.empty())
-        {
             wrlock_owner = writers_queue.begin();
-        }
         else
-        {
             rdlock_owner = readers_queue.begin();
-        }
     }
     else
     {
         writers_queue.erase(group_it);
         /// Prepare next phase
         if (!readers_queue.empty())
-        {
-            if (next && readers_queue.size() > 1)
-            {
-                rdlock_owner = std::next(readers_queue.begin());
-            }
-            else
-            {
-                rdlock_owner = readers_queue.begin();
-            }
-        }
+            rdlock_owner = readers_queue.begin();
         else
-        {
             wrlock_owner = writers_queue.begin();
-        }
     }
 
     if (rdlock_owner != readers_queue.end())
     {
-        rdlock_owner->cv.notify_all();
+        grantOwnershipToAllReaders();
     }
     else if (wrlock_owner != writers_queue.end())
     {
-        wrlock_owner->cv.notify_one();
+        grantOwnership(wrlock_owner);
+    }
+}
+
+
+void RWLockImpl::grantOwnership(GroupsContainer::iterator group_it) noexcept
+{
+    if (!group_it->ownership)
+    {
+        group_it->ownership = true;
+        group_it->cv.notify_all();
+    }
+}
+
+
+void RWLockImpl::grantOwnershipToAllReaders() noexcept
+{
+    if (rdlock_owner != readers_queue.end())
+    {
+        for (;;)
```

Comment:
There should be, at most, only one group of readers pending right?

You can have N readers that are already owner, a writers, and then only a block of readers all together. AFAICS it's not possible to have more than one group of readers that aren't owners, but I might be missing something

Reply:
Yes, we append readers to the same reader group if this group doesn't have ownership yet.
But there can be multiple reader groups with ownership.

Reply:
I've added `chassert()` to check that.